### PR TITLE
Add support for 'ninja' and for only using one process for linking

### PR DESCRIPTION
**Pull request by andrewvaughanj**
## Issue

Over the weekend, I was trying to build `facebook-clang-plugins` on quite beefy machine (24 GiB of RAM, 12 GiB of swap, 12 core Xeon @ 3.60 GHz), however I _still_ got the dreaded OOM killer when linking (when more than one link tries to happen simultaneously).

## Resolution

This PR introduces two *optional* changes:

1. It allows a user to use [`ninja`](https://ninja-build.org) to build rather than GNU Make (to quote the Ninja page "Ninja is used to build Google Chrome, parts of Android, LLVM")

2. It allows the user to [restrict linking to only one parallel job](https://gitlab.kitware.com/cmake/cmake/commit/4a1f996d47e69b17276c2773c44036711b9376fa) by using Ninja [pools](https://ninja-build.org/manual.html#ref_pool) to avoid many parallel links -- sequential linking is only allowed when using Ninja

If this PR is acceptable, I would be interested in changing Infer's `build_infer.sh` to mirror these arguments, such that a user can run `./build_infer.sh -s -n clang` and this will build `facebook-clang-plugins` using Ninja and with sequential linking.

## Questions

When using Ninja, it is not really "best practise" to control the number of parallel jobs; this is up to Ninja to decide (typically, it uses [CPUs + 2](http://www.linuxfromscratch.org/lfs/view/development/chapter06/ninja.html)).

However, right now there are two things we might want to change:

- [ ] When using Ninja, it is not possible for the user to limit the number of jobs (i.e., we do not set `BUILD_ARGS` to contain a `-j` for Ninja`) -- is this a problem?

- [ ] When building with sequential linking, the number of Ninja jobs _for compilation_ is set to `NCPUS`, which is different to `JOBS` (which is half this number) and also different to the Ninja default (which would be NCPUS + 2) -- is this a problem?

For my usage, neither of these are a problem, but if they are, then I can:

1. Reintroduce a `-j` for Ninja (set to $NCPUS + 2) and then the user can set `_NPROCESSORS_ONLN` themselves

2. Calculate `NCPUS + 2` for controlling the Ninja pools (again, the user can set `_NPROCESSORS_ONLN` themselves to control this total count)

Thoughts?

Signed-off-by: Andrew V. Jones <andrewvaughanj@gmail.com>

### CODE REVIEW

File: clang/setup.sh [hunk 1]
```diff
@@ -35,6 +35,8 @@ usage () {
     echo "    -c,--only-check-install    check if recompiling clang is needed"
     echo "    -h,--help                  show this message"
     echo "    -r,--only-record-install   do not install clang but pretend we did"
+    echo "    -n|--ninja                 use ninja for building"
+    echo "    -s|--sequential-link       only use one process for linking (ninja only)"
```
Comment (jvillard) [inline clang/setup.sh:39]: please follow the same still as the other options: `-n,--ninja` and `-s,--sequential-link`.

Comment (andrewvaughanj) [reply clang/setup.sh:39]: Oops! That's definitely a `yyp` kind of error :p

Comment (jvillard) [reply clang/setup.sh:39]: Also we should probably keep these sorted alphabetically.

Comment (andrewvaughanj) [reply clang/setup.sh]: Both done!

**Comment by jvillard** [comment]
Many thanks! :rocket:

**Pull request merged**